src/manifolds/Segre.jl

```diff
@@ -0,0 +1,473 @@
+````
+is the set of rank-one tensors in ``𝔽^{n_1} \otimes \dots \otimes 𝔽^{n_d}``.
+
+When ``𝔽 = ℝ``, the Segre manifold is represented as
+````math
+    \mathcal{S} \sim ℝ^{+} \times S^{n_1 - 1} \times \dots \times S^{n_d - 1}.
+````
+This is a local diffeomorphism, and the metric is a locally a [warped product](https://en.wikipedia.org/wiki/Warped_product) of ``ℝ^{+}`` and ``S^{n_1 - 1} \times \dots \times S^{n_d - 1}``. The tuple ``(n_1, \dots, n_d)`` is called the _valence_ of the manifold.
+
+The geometry is summarized in [JacobssonSwijsenVandervekenVannieuwenhoven:2024](@cite).
+
+# Constructor
+    Segre(valence::NTuple{V, Int}; field::AbstractNumbers=ℝ)
+
+Generate a valence `V` Segre manifold.
+"""
+struct Segre{𝔽,V} <: AbstractManifold{𝔽} end
+
+function Segre(valence::NTuple{V,Int}; field::AbstractNumbers=ℝ) where {V}
+    return Segre{field,valence}()
+end
+
+valence(::Segre{𝔽,V}) where {𝔽,V} = V
+ndims(::Segre{𝔽,V}) where {𝔽,V} = length(V)
+manifold_dimension(::Segre{𝔽,V}) where {𝔽,V} = (1 + sum(V .- 1))
+
+"""
+    check_size(M::Segre{𝔽, V}, p)
+
+Check whether `p` has the right size for `Segre` manifold `M`.
+"""
+function check_size(M::Segre{𝔽,V}, p;) where {𝔽,V}
+    p_size = only.(size.(p))
+    M_size = [1, V...]
+
+    if p_size != M_size
+        return DomainError(
+            p_size,
+            "The point $(p) can not belong to the manifold $(M), since its size $(p_size) is not equal to the manifolds representation size ($(M_size)).",
+        )
+    end
+
+    return nothing
+end
+
+"""
+    check_size(M::Segre{𝔽, V}, p, v)
+
+Check whether `p` and `v` have the right size for `Segre` manifold `M`.
+"""
+function check_size(M::Segre{𝔽,V}, p, v;) where {𝔽,V}
+    p_size = only.(size.(p))
+    v_size = only.(size.(v))
+    M_size = [1, V...]
+
+    if p_size != M_size
+        return DomainError(
+            p_size,
+            "The point $(p) can not belong to the manifold $(M), since its size $(p_size) is not equal to the manifolds representation size ($(M_size)).",
+        )
+    end
+
+    if v_size != M_size
+        return DomainError(
+            v_size,
+            "The vector $(v) can not belong to the manifold $(M), since its size $(v_size) is not equal to the manifolds representation size ($(M_size)).",
+        )
+    end
+
+    return nothing
+end
+
+"""
+    check_point(M::Segre{ℝ, V}, p; kwargs...)
+
+Check whether `p` is a valid point on `M`, i.e. `p[1]` is a singleton containing a positive number and `p[i + 1]` is a point on `Sphere(V[i])`. The tolerance can be set using the `kwargs...`.
+"""
+function check_point(M::Segre{ℝ,V}, p; kwargs...) where {V}
+    if p[1][1] <= 0.0
+        return DomainError(p[1][1], "$(p) has non-positive modulus.")
+    end
+
+    for (x, n) in zip(p[2:end], V)
+        e = check_point(Sphere(n - 1)::AbstractSphere, x; rtol=1e-10, atol=1e-10, kwargs...)
+        if !isnothing(e)
+            return e
+        end
+    end
```

Comment:
(besides writing a test for this as well...)

These checks currently have a small disadvantage. if the components of `p` that are checked here have several mistakes, the user would always only see the first one.
After fixing that he would see the second,...

In the product manifold, we for example “collect” these and have created a special error message type to report all such errors, cf. https://github.com/JuliaManifolds/ManifoldsBase.jl/blob/76d6a5fb05a1593325dd342ae780dd89f55afbcd/src/ProductManifold.jl#L169-L175 and the `ComponentManifoldError` that contains a tuple of errors. Would that maybe make sense here as well?

Reply:
Neat idea indeed! You have to switch it, first M then p, but the rest looks good :)